pkg/apis/core/types.go

```diff
@@ -5788,3 +5796,470 @@ type PortStatus struct {
+	metav1.TypeMeta
+
+	// The driver must set a finalizer here before it attempts to allocate
+	// the resource. It removes the finalizer again when a) the allocation
+	// attempt has definitely failed or b) when the allocated resource was
+	// freed. This ensures that resources are not leaked.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta
+
+	// Spec describes the desired attributes of a resource that then needs
+	// to be allocated. It can only be set once when creating the
+	// ResourceClaim.
+	Spec ResourceClaimSpec
+
+	// Status describes whether the resource is available and with which
+	// attributes.
+	Status ResourceClaimStatus
+}
+
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ResourceClaimStatus struct {
+	// DriverName is a copy of the driver name from the ResourceClass at
+	// the time when allocation started. It's necessary to enable usage of
+	// the claim by kubelet in case that the ResourceClass got removed in
+	// the meantime. It also helps the resource driver to determine whether
+	// it needs to handle a claim that got marked for deletion.
+	DriverName string
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. Nil indicates that the resource is not
+	// allocated.
+	Allocation *AllocationResult
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. Usually those are Pods, but other objects are also
+	// possible as long as they exist as resources in the apiserver. The
+	// ResourceClaim controller in the kube-controller-manager checks for
+	// stale entries that reference non-existent objects and removes them.
+	//
+	// A scheduler must add a Pod that it is scheduling. This must be done
+	// in an atomic ResourceClaim update because there might be multiple
+	// schedulers working on different Pods that compete for access to the
+	// same ResourceClaim, the ResourceClaim might have been marked for
+	// deletion, or even been deallocated already.
+	//
+	// kubelet will check this before allowing a Pod to run because a a
+	// user might have selected a node manually without reserving resources
+	// or a scheduler might have missed that step, for example because it
+	// doesn't support dynamic resource allocation or the feature was
+	// disabled.
+	//
+	// The maximum size is 32 (= [ResourceClaimReservedForMaxSize]). This
+	// is an artificial limit to prevent a completely unbounded field in
+	// the API.
+	ReservedFor []ResourceClaimUserReference
+
+	// DeallocationRequested gets set by the scheduler when it detects the
+	// situation where pod scheduling cannot proceed because some claim was
+	// allocated for a node that cannot provide some other required
+	// resource.
+	//
+	// The driver then needs to deallocate this claim and the scheduler
+	// will try again.
```

Comment:
Is the driver expected to set it back to false? Does ReservedFor need to be empty when setting this to true?

Reply:
> Is the driver expected to set it back to false?

Yes. I added an explanation regarding that.

> Does ReservedFor need to be empty when setting this to true?

It wasn't required (deallocation would have been blocked while reserved), but it makes sense to add that because the scheduler should pick some other claim for deallocation or wait. Added.